### ISSUE TITLE
[Feature Request] 

### ISSUE THREAD

**Issue by MuhammadSawalhy** [jesseduffield/lazygit#1782]
**Is your feature request related to a problem? Please describe.**

1. I have some staged changes that I want to keep and discard the unstaged changes including unstaged untracked files.
2. Or do it otherwise, discard the staged and keep the unstaged changes unstaged.

**Describe the solution you'd like**

In the discard menu options which opens by pressing <kbd>d</kbd> we should find these options.

**Describe alternatives you've considered**

1. `git stash save --keep-index --include-untracked && git stash drop` or simply `git restore .`
2. `patch -p1 -R < <(git diff --cached) && git reset`

**Additional context**

![image](https://user-images.githubusercontent.com/42011920/154963025-9751c979-92ee-4536-af97-4b4813ae487f.png)

**Comment by jesseduffield** [comment]
@MuhammadSawalhy that code snippet you posted looks a little gnarly. Do you know of any other way to achieve the same result?

**Comment by MuhammadSawalhy** [comment]
> Do you know of any other way to achieve the same result?

I searched google for help and got this [StackOverflow answer](https://stackoverflow.com/a/56221568/10891757)

```
git stash save --keep-index
git reset --hard
git stash pop
```

**Comment by MuhammadSawalhy** [comment]
In my case the above solution doesn't work for me, I just copy-pasted it here without trying it. So I think we can go with another trick

```
git commit -m "temp commit"
git stash
git reset --hard HEAD^
git stash pop
```

**Comment by mark2185** [comment]
We already have the `stash only staged changes` provided by #1870 so this could technically be done by:
1. stash only staged changes
2. drop stash

@jesseduffield I could write it up and open a PR if that sounds like a reasonable approach, what do you think?

**Comment by jesseduffield** [comment]
Sounds good to me @mark2185

**Comment by MuhammadSawalhy** [comment]
With my poor connection, it seems that I am cloning tons of files 😅 

![image](https://user-images.githubusercontent.com/42011920/173453325-f47d8d6e-b95a-41a3-bb09-8298b141d141.png)

**Comment by MuhammadSawalhy** [comment]
Discarding staged changes only is working 🎉, but there is no option to discard only staged changes of a specific file:

![image](https://user-images.githubusercontent.com/42011920/173457248-824d1c75-2669-4b20-b9c6-e9a1564bd029.png)

**Comment by mark2185** [comment]
You're right, if a file is partially staged we currently offer to discard either _everything_ or _just the unstaged changes_.

I'll take a look into that and get back to you.

**Comment by izelnakri** [comment]
Hi there, can we prioritize this issue? I would like to be able to "Drop Staged changes" only, without having to drop everything that is unstaged for a single file.

Right now I cant "drop a staged hunk/selection"s with a single keypress as well, first those hunks/selections have to be unstaged, then I need to reselect them and can discard. This flow needs to improve significantly.

**Comment by jesseduffield** [comment]
@izelnakri it's not clear to me how to implement either of those features: if you would like to do some digging and propose a solution using git CLI commands, or find another open source git UI that supports the feature, that will help move this issue along.

Specifically regarding dropping a staged hunk, that sounds especially tricky because the hunk's context often changes upon being unstaged, so if we tried removing the patch in two steps: first unstaging it, then removing it, it may simply not work.

**Comment by izelnakri** [comment]
Hi @jesseduffield first of all, thanks alot for this amazing tool! Its been a productivity boost, I don't have to tmux switch panes and run most of my git commands anymore.

Secondly, I've seen a tendency of declaring certain issues as "technically impossible". Maybe thats a coincidence I've seen a similar thing here, and in both cases issues are technically possible: https://github.com/jesseduffield/lazygit/issues/3297#issuecomment-1923372937 this issue is actually ongoing on https://github.com/jesseduffield/lazygit/issues/2117

I know that these issues are technically possible because I've been using them on Vimagit for the last 6 years: https://github.com/jreybert/vimagit . I'll have a lookg how they implement it and describe it here if needed. I don't write in go language usually so I'm not very comfortable with it although understand most/all of it almost. I  might also able to take a stab at it, probably in the next couple of months due to my current workload.
